apps/emqx/test/emqx_persistent_session_ds_fuzzer.erl

```diff
@@ -0,0 +1,738 @@
+connect(S, Opts = #{clientid := ClientId}) ->
+    ?tp(info, ?sessds_test_connect, #{opts => Opts, pid => self()}),
+    %% Check metadata of the previous state to catch situations when
+    %% the testcase starts from a dirty state:
+    true = check_session_metadata(S),
+    %% Start the new client:
+    {ok, ClientPid} = emqtt:start_link(Opts),
+    unlink(ClientPid),
+    {ok, _} = emqtt:connect(ClientPid),
+    %% Wait for the old client stepdown:
+    maybe_wait_stepdown(S),
+    %% Update connection info:
+    register(?client, ClientPid),
+    [SessionPid] = emqx_cm:lookup_channels(local, ClientId),
+    %% Return `conninfo()':
+    #{
+        client_pid => ClientPid,
+        session_pid => SessionPid
+    }.
+
+%% @doc Shut down emqtt
+disconnect(#{conninfo := ConnInfo = #{client_pid := CPid}}) ->
+    ?tp(info, ?sessds_test_disconnect, #{pid => CPid}),
+    emqtt:stop(client_pid()),
+    wait_stepdown(ConnInfo),
+    flush_client_messages(CPid).
+
+publish(Batch) ->
+    %% Produce traces for each message we're about to publish:
+    [?tp(info, ?sessds_test_out_publish, emqx_message:to_map(Msg)) || Msg <- Batch],
+    %% We bypass persistent session router for simplicity:
+    ok = emqx_ds:store_batch(?PERSISTENT_MESSAGE_DB, [
+        Msg#message{timestamp = emqx_message:timestamp_now()}
+     || Msg <- Batch
+    ]).
+
+add_generation() ->
+    ?tp(info, ?sessds_test_add_generation, #{}),
+    emqx_ds:add_generation(?PERSISTENT_MESSAGE_DB).
+
+subscribe(Topic, QoS) ->
+    ?tp(info, ?sessds_test_subscribe, #{topic => Topic, qos => QoS}),
+    emqtt:subscribe(client_pid(), Topic, QoS).
+
+unsubscribe(Topic) ->
+    ?tp(info, ?sessds_test_unsubscribe, #{topic => Topic}),
+    emqtt:unsubscribe(client_pid(), Topic).
+
+consume(S = #{conninfo := #{client_pid := CPID, session_pid := SPID}}) ->
+    %% Set up monitoring to detect crashes early:
+    %% Consume and ack all messages we can get from the broker:
+    CMRef = monitor(process, CPID),
+    SMRef = monitor(process, SPID),
+    ?tp_span(
+        info,
+        ?sessds_test_consume,
+        #{},
+        receive_ack_loop(S, CMRef, SMRef, ok)
+    ).
+
+%% @doc This function receives and acknowledges all MQTT messages sent
+%% to the client
+receive_ack_loop(
+    S = #{conninfo := #{client_pid := CPID}},
+    CMRef,
+    SMRef,
+    Result
+) ->
+    receive
+        %% Handle MQTT packets:
+        {publish, Msg = #{client_pid := CPID}} ->
+            ?tp(info, ?sessds_test_in_publish, Msg),
+            #{packet_id := PID, qos := QoS} = Msg,
+            %% Ack:
+            case QoS of
+                ?QOS_0 ->
+                    ok;
+                ?QOS_1 ->
+                    ?tp(info, ?sessds_test_out_puback, #{packet_id => PID}),
+                    emqtt:puback(client_pid(), PID);
+                ?QOS_2 ->
+                    ?tp(info, ?sessds_test_out_pubrec, #{packet_id => PID}),
+                    emqtt:pubrec(client_pid(), PID)
```

Comment:
Maybe occasionally replying with the wrong ack (i.e., puback for qos =/= 1, pubrec/pubcomp for qos =/= 2) might help simulate [rogue/buggy clients](https://github.com/emqx/emqx/issues/10215)?

Reply:
Yes, it should be possible to extend the fuzzer, but I'd prefer to do it in a later PR.